rllib/utils/replay_buffers/episode_replay_buffer.py

```diff
@@ -0,0 +1,508 @@
+        self._indices = []
+
+        # The size of the buffer in timesteps.
+        self.size = 0
+        # How many timesteps have been sampled from the buffer in total?
+        self.sampled_timesteps = 0
+
+    @override(ReplayBufferInterface)
+    def __len__(self) -> int:
+        return self.get_num_timesteps()
+
+    @override(ReplayBufferInterface)
+    def add(self, batch: SampleBatchType, **kwargs) -> None:
+        """Converts the incoming SampleBatch into a number of _Episode objects.
+
+        Then adds these episodes to the internal deque.
+        """
+        if isinstance(batch, MultiAgentBatch):
+            raise ValueError(
+                "`EpisodeReplayBuffer` cannot operate on MultiAgentBatches yet! "
+                "For single-agent use only."
+            )
+
+        episode_slices = batch.split_by_episode()
+        episodes = [
+            _Episode.from_sample_batch(eps_slice)
+            for eps_slice in episode_slices
+        ]
+
+        for eps in episodes:
+            self.size += len(eps)
+            # Ongoing episode, concat to existing record.
+            if eps.id_ in self.episode_id_to_index:
+                buf_idx = self.episode_id_to_index[eps.id_]
+                existing_eps = self.episodes[buf_idx - self._num_episodes_ejected]
+                old_len = len(existing_eps)
+                self._indices.extend([(buf_idx, old_len + i) for i in range(len(eps))])
+                existing_eps.concat_episode(eps)
+            # New episode. Add to end of our episodes deque.
+            else:
+                self.episodes.append(eps)
+                buf_idx = len(self.episodes) - 1 + self._num_episodes_ejected
+                self.episode_id_to_index[eps.id_] = buf_idx
+                self._indices.extend([(buf_idx, i) for i in range(len(eps))])
+
+            # Eject old records from front of deque.
+            while self.size > self.capacity:
+                # Eject oldest episode.
+                ejected_eps = self.episodes.popleft()
+                ejected_eps_len = len(ejected_eps)
+                # Correct our size.
+                self.size -= len(ejected_eps)
+                # Erase episode from all our indices.
+                # Main episode index.
+                ejected_idx = self.episode_id_to_index[ejected_eps.id_]
+                del self.episode_id_to_index[ejected_eps.id_]
+
+                # All timestep indices that this episode owned.
+                new_indices = []  # New indices that will replace self._indices.
+                idx_cursor = 0
+                for i, idx_tuple in enumerate(self._indices):
+                    if idx_cursor is not None and idx_tuple[0] == ejected_idx:
+                        new_indices.extend(self._indices[idx_cursor:i])
+                        idx_cursor = None
+                    elif idx_cursor is None:
+                        if idx_tuple[0] != ejected_idx:
+                            idx_cursor = i
```

Comment:
Wouldn't running into this condition `idx_tuple[0] != ejected_idx` mean that the buffer or the episode object is broken? And if so, would we simply hide that fact here?
Reasoning would be that if we length of the episode that we eject is correct, `idx_tuple[1] == ejected_eps_len - 1` should always kick in one step before `idx_tuple[0] != ejected_idx`.

Reply:
I explained this better now via more comments.

There are two phases that the looping through `self._indices` can be in:
1) i is pointing to a tuple (eps_idx, ts_in_eps_idx) that's inside a non-evicted episode.
2) i is pointing to a tuple that's inside the just evicted episode.

When 1): idx_cursor is an int pointing to the first entry to keep for the new compiled `self._indices`.
When 2): idx_cursor is None
When transitioning from 1) to 2): collect all tuples from `idx_cursor` to i into the new compiled `self._indices` to make sure the non-evicted episodes will be in there.
When transitioning from 2) to 1): Do NOT anything to the new compiled `self._indices` (in order to skip the evicted episode) and set `idx_cursor` to the current i.

Reply:
I went through the logic again and think that it's correct. But I will add a proper test case for the eviction logic ...

Reply:
Got it, thanks!